### ISSUE TITLE
dimensions in forward pass

### ISSUE THREAD

**Issue by soldierofhell** [zhiqwang/image-captioning#16]
@zhiqwang, could you please correct dimensions I've added below in comments in the forward() pass of CRNN class, because I cannot figure out what happens after permute line
```
out = self.features(x) # out: (B, H, W, C)
# features -> pool -> flatten -> decoder -> softmax
out = self.avgpool(out) # out: (B, 1, W, C)
out = out.permute(3, 0, 1, 2).view(out.size(3), out.size(0), -1) # out: (C, B, 1*W)
out = self.classifier(out) # expected in: (B, W, C) != (C, B, 1*W) ?
```

**Comment by zhiqwang** [comment]
Hi @soldierofhell 
```
out = self.features(x)  # out: (B, C, H, W), W for `seq_len`
# where C is channel of CNN backbone, for example 342 if `densenet_cifar`.
# features -> pool -> flatten -> decoder -> softmax
out = self.avgpool(out)  # out: (B, C, 1, W)
out = out.permute(3, 0, 1, 2).view(out.size(3), out.size(0), -1)  # out: (W, B, 1*C)
out = self.classifier(out)  # expected in: (W, B, vocab)
# where vocab is the number of alphabets plus one (for blank token)

**Comment by soldierofhell** [comment]
Thanks, but still don't get why `in:(B, 1, W, C) -> .permute(3, 0, 1, 2).view(.size(3), .size(0), -1) -> out:(W, B, 1*C)`
Seems like it should be `permute(2, 0, 1, 3).view()`?

**Comment by zhiqwang** [comment]
Sorry @soldierofhell , I modify the dimension, it is (B, C, H, W) in the first line.

And H, W in the first line is not the size of input images, it is the down-sampled size passing the CNN backbone, for `densenet_cifar`, it compress image's width by 1/4. refer issue #4 .

**Comment by soldierofhell** [comment]
Ok, now everything is clear. It's all because I'm used to tensorflow channel last convention :)
